File: src/main/kotlin/com/kursx/parser/fb2/FictionBook.java
package com.kursx.parser.fb2;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jsoup.Jsoup;
import org.jsoup.helper.W3CDom;
import org.jsoup.parser.Parser;
import org.tinylog.Logger;
import org.tinylog.TaggedLogger;
import org.w3c.dom.Document;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;
import org.xml.sax.SAXParseException;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.stream.XMLInputFactory;
import javax.xml.stream.XMLStreamException;
import java.io.*;
import java.util.*;

public class FictionBook {

    public static final TaggedLogger FB2_LOGGER = Logger.tag("FB2");
    protected Xmlns[] xmlns;
    protected Description description;
    protected List<Body> bodies = new ArrayList<>();
    protected Map<String, Binary> binaries;

    public String encoding = "utf-8";

    public FictionBook() {
    }

    public FictionBook(File file) throws XMLStreamException, IOException, ParserConfigurationException {
        Document doc;
        try {
            String inferred = XMLInputFactory.newInstance().createXMLStreamReader(new FileReader(file)).getCharacterEncodingScheme();
            encoding = inferred == null ? encoding : inferred;

            DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
            DocumentBuilder db = dbf.newDocumentBuilder();
            InputStream inputStream = new FileInputStream(file);
            BufferedReader br = new BufferedReader(new FileReader(file));
            boolean foundIllegalCharacters = false;
            try {
                StringBuilder line = new StringBuilder(br.readLine().trim());
                if (!line.toString().startsWith("<")) {
                    foundIllegalCharacters = true;
                }
            } catch (Exception e) {
                FB2_LOGGER.error(e);
            }
            if (foundIllegalCharacters) {
                FB2_LOGGER.debug(() -> "Found illegal characters in " + file.getAbsolutePath());
                StringBuilder text = new StringBuilder();
                br = new BufferedReader(new FileReader(file));
                String line = br.readLine();
                if (line != null && line.contains("<")) {
                    line = line.substring(line.indexOf("<"));
                }
                while (line != null) {
                    text.append(line);
                    line = br.readLine();
                }
                br.close();
                try {
                    FB2_LOGGER.debug(() -> "Parsing " + file.getAbsolutePath() + " without explicitly specified encoding");
                    doc = db.parse(new InputSource(new StringReader(text.toString())));
                } catch (SAXException sax) {
                    FB2_LOGGER.warn(sax, () -> "Parsing " + file.getAbsolutePath() + " filed, falling back to Jsoup");
                    doc = W3CDom.convert(Jsoup.parse(file, encoding, "https://yandex.ru", Parser.xmlParser()));
                }
            } else {
                try {
                    FB2_LOGGER.debug(() -> "Parsing " + file.getAbsolutePath() + " with encoding " + encoding);
                    doc = db.parse(new InputSource(new InputStreamReader(inputStream, encoding)));
                } catch (SAXException sax) {
                    FB2_LOGGER.warn(sax, () -> "Parsing " + file.getAbsolutePath() + " filed, falling back to Jsoup");
                    doc = W3CDom.convert(Jsoup.parse(file, encoding, "https://yandex.ru", Parser.xmlParser()));
                }
            }
        } catch (IOException | ParserConfigurationException | XMLStreamException e) {
            FB2_LOGGER.error(e, () -> "Error while processing " + file.getAbsolutePath());
            throw e;
        }
        initXmlns(doc);
        description = new Description(doc);
        NodeList bodyNodes = doc.getElementsByTagName("body");
        for (int item = 0; item < bodyNodes.getLength(); item++) {
            bodies.add(new Body(bodyNodes.item(item)));
        }
        NodeList binary = doc.getElementsByTagName("binary");
        for (int item = 0; item < binary.getLength(); item++) {
            if (binaries == null) binaries = new HashMap<>();
            Binary binary1 = new Binary(binary.item(item));
            try {
                binaries.put(binary1.getId().replace("#", ""), binary1);
            } catch (Exception e) {
                FB2_LOGGER.error(e, () -> "Invalid binary " + binary1 + " in file " + file.getAbsolutePath());
            }
        }
    }

    protected void setXmlns(ArrayList<Node> nodeList) {
        xmlns = new Xmlns[nodeList.size()];
        for (int index = 0; index < nodeList.size(); index++) {
            Node node = nodeList.get(index);
            xmlns[index] = new Xmlns(node);
        }
    }

    protected void initXmlns(Document doc) {
        NodeList fictionBook = doc.getElementsByTagName("FictionBook");
        ArrayList<Node> xmlns = new ArrayList<>();
        for (int item = 0; item < fictionBook.getLength(); item++) {
            NamedNodeMap map = fictionBook.item(item).getAttributes();
            for (int index = 0; index < map.getLength(); index++) {
                Node node = map.item(index);
                xmlns.add(node);
            }
        }
        setXmlns(xmlns);
    }

    public ArrayList<Person> getAuthors() {
        return description.getDocumentInfo().getAuthors();
    }

    public Xmlns[] getXmlns() {
        return xmlns;
    }

    public Description getDescription() {
        return description;
    }

    public @Nullable Body getBody() {
        return getBody(null);
    }

    public @Nullable Body getNotes() {
        return getBody("notes");
    }

    public @Nullable Body getComments() {
        return getBody("comments");
    }

    private @NotNull Body getBody(String name) {
        for (Body body : bodies) {
            if ((name + "").equals(body.getName() + "")) {
                return body;
            }
        }
        return bodies.get(0);
    }

    @NotNull
    public Map<String, Binary> getBinaries() {
        return binaries == null ? new HashMap<String, Binary>() : binaries;
    }

    public String getTitle() {
        return description.getTitleInfo().getBookTitle();
    }

    public String getLang() {
        return description.getTitleInfo().getLang();
    }

    public @Nullable Annotation getAnnotation() {
        return description.getTitleInfo().getAnnotation();
    }
}

File: src/main/kotlin/com/kursx/parser/fb2/FictionBook.kt
package com.kursx.parser.fb2

import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import kotlinx.serialization.protobuf.ProtoNumber
import net.lingala.zip4j.ZipFile
import net.lingala.zip4j.model.FileHeader
import org.jsoup.Jsoup
import org.jsoup.helper.W3CDom
import org.jsoup.parser.Parser
import org.tinylog.Logger
import org.tinylog.TaggedLogger
import org.w3c.dom.Document
import org.w3c.dom.Node
import org.xml.sax.InputSource
import org.xml.sax.SAXException
import java.io.*
import javax.xml.parsers.DocumentBuilder
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.parsers.ParserConfigurationException
import javax.xml.stream.XMLInputFactory
import javax.xml.stream.XMLStreamException

@Serializable
class FictionBook {
    @ProtoNumber(1)
    private var xmlns: Array<Xmlns?> = Array(0){null}
    @ProtoNumber(2)
    var description: Description? = null
        private set
    @Transient
    private var bodies: MutableList<Body> = ArrayList()
    @ProtoNumber(3)
    var binaries: MutableMap<String, Binary> = hashMapOf()
    @ProtoNumber(4)
    private var encoding = "utf-8"

    constructor(zipFile: ZipFile, fileHeader: FileHeader) {
        fun isp() = zipFile.getInputStream(fileHeader)
        val zipFilePath = zipFile.file.absolutePath + "#" + fileHeader.fileName
        val doc: Document
        try {
            encoding = zipFile.getInputStream(fileHeader).use { inp ->
                XMLInputFactory.newInstance().createXMLStreamReader(inp).characterEncodingScheme
            } ?: encoding
            val dbf = DocumentBuilderFactory.newInstance()
            val db = dbf.newDocumentBuilder()
            var foundIllegalCharacters = false
            BufferedReader(InputStreamReader(isp())).use { br ->
                try {
                    val line = StringBuilder(br.readLine().trim { it <= ' ' })
                    if (!line.toString().startsWith("<")) {
                        foundIllegalCharacters = true
                    }
                } catch (e: Exception) {
                    FB2_LOGGER.error(e)
                }
            }
            doc = if (foundIllegalCharacters) {
                FB2_LOGGER.debug { "Found illegal characters in $zipFilePath" }
                val text = StringBuilder()
                text.fillFromReader(BufferedReader(InputStreamReader(isp())))
                try {
                    FB2_LOGGER.debug { "  Parsing $zipFilePath without explicitly specified encoding" }
                    db.parse(InputSource(StringReader(text.toString())))
                } catch (sax: SAXException) {
                    FB2_LOGGER.warn(sax) { "    Parsing $zipFilePath filed, falling back to Jsoup" }
                    BufferedInputStream(isp()).use { br ->
                        W3CDom.convert(Jsoup.parse(br, encoding, "https://yandex.ru", Parser.xmlParser()))
                    }
                }
            } else {
                try {
                    FB2_LOGGER.debug { "Parsing $zipFilePath with encoding $encoding" }
                    db.fromEncodedInputStream(isp())
                } catch (sax: SAXException) {
                    FB2_LOGGER.warn(sax) { "  Parsing $zipFilePath filed, falling back to Jsoup" }
                    BufferedInputStream(isp()).use { br ->
                        W3CDom.convert(Jsoup.parse(br, encoding, "https://yandex.ru", Parser.xmlParser()))
                    }
                }
            }
        } catch (e: IOException) {
            FB2_LOGGER.error(e) { "Error while processing $zipFilePath" }
            throw e
        } catch (e: ParserConfigurationException) {
            FB2_LOGGER.error(e) { "Error while processing $zipFilePath" }
            throw e
        } catch (e: XMLStreamException) {
            FB2_LOGGER.error(e) { "Error while processing $zipFilePath" }
            throw e
        }
        afterInit(doc, zipFilePath)
    }

    constructor(file: File) {
        val doc: Document
        try {
            encoding = FileReader(file).use { reader ->
                XMLInputFactory.newInstance().createXMLStreamReader(reader).characterEncodingScheme
            } ?: encoding
            val dbf = DocumentBuilderFactory.newInstance()
            val db = dbf.newDocumentBuilder()
            var foundIllegalCharacters = false
            BufferedReader(FileReader(file)).use { br ->
                try {
                    val line = StringBuilder(br.readLine().trim { it <= ' ' })
                    if (!line.toString().startsWith("<")) {
                        foundIllegalCharacters = true
                    }
                } catch (e: Exception) {
                    FB2_LOGGER.error(e)
                }
            }
            doc = if (foundIllegalCharacters) {
                FB2_LOGGER.debug { "Found illegal characters in ${file.absolutePath}" }
                val text = StringBuilder()
                text.fillFromReader(BufferedReader(FileReader(file)))
                try {
                    FB2_LOGGER.debug { "Parsing ${file.absolutePath} without explicitly specified encoding" }
                    db.parse(InputSource(StringReader(text.toString())))
                } catch (sax: SAXException) {
                    FB2_LOGGER.warn(sax) { "Parsing ${file.absolutePath} filed, falling back to Jsoup" }
                    W3CDom.convert(Jsoup.parse(file, encoding, "https://yandex.ru", Parser.xmlParser()))
                }
            } else {
                try {
                    FB2_LOGGER.debug { "Parsing ${file.absolutePath} with encoding $encoding" }
                    db.fromEncodedInputStream(FileInputStream(file))
                } catch (sax: SAXException) {
                    FB2_LOGGER.warn(sax) { "Parsing ${file.absolutePath} filed, falling back to Jsoup" }
                    W3CDom.convert(Jsoup.parse(file, encoding, "https://yandex.ru", Parser.xmlParser()))
                }
            }
        } catch (e: Exception) {
            when (e) {
                is IOException, is ParserConfigurationException, is XMLStreamException ->
                    FB2_LOGGER.error(e) { "Error while processing ${file.absolutePath}" }
            }
            throw e
        }
        afterInit(doc, file.absolutePath)
    }

    private fun StringBuilder.fillFromReader(reader: BufferedReader) = reader.use { br ->
        var line = br.readLine()
        if (line != null && line.contains("<")) {
            line = line.substring(line.indexOf("<"))
        }
        while (line != null) {
            append(line)
            line = br.readLine()
        }
    }

    private fun DocumentBuilder.fromEncodedInputStream(inputStream: InputStream): Document {
        return InputStreamReader(BufferedInputStream(inputStream), encoding).use { br ->
            parse(InputSource(br))
        }
    }

    private fun afterInit(doc: Document, path: String) {
        initXmlns(doc)
        description = Description(doc)
        val bodyNodes = doc.getElementsByTagName("body")
        for (item in 0 until bodyNodes.length) {
            bodies.add(Body(bodyNodes.item(item)))
        }
        val binary = doc.getElementsByTagName("binary")
        for (item in 0 until binary.length) {
            val binary1 = Binary(binary.item(item))
            try {
                binaries[binary1.id!!.replace("#", "")] = binary1
            } catch (e: Exception) {
                FB2_LOGGER.error(e) { "Invalid binary $binary1 in file $path" }
            }
        }
    }

    private fun setXmlns(nodeList: ArrayList<Node?>) {
        xmlns = arrayOfNulls(nodeList.size)
        for (index in nodeList.indices) {
            val node = nodeList[index]!!
            xmlns[index] = Xmlns(node)
        }
    }

    private fun initXmlns(doc: Document) {
        val fictionBook = doc.getElementsByTagName("FictionBook")
        val xmlns = ArrayList<Node?>()
        for (item in 0 until fictionBook.length) {
            val map = fictionBook.item(item).attributes
            for (index in 0 until map.length) {
                val node = map.item(index)
                xmlns.add(node)
            }
        }
        setXmlns(xmlns)
    }

    val authors: ArrayList<Person>
        get() = description!!.documentInfo!!.authors
    val body: Body
        get() = getBody(null)
    val notes: Body
        get() = getBody("notes")
    val comments: Body
        get() = getBody("comments")

    private fun getBody(name: String?): Body {
        for (body in bodies) {
            if (name == body.name) {
                return body
            }
        }
        return bodies[0]
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as FictionBook

        if (!xmlns.contentEquals(other.xmlns)) return false
        if (description != other.description) return false
        if (binaries != other.binaries) return false
        if (encoding != other.encoding) return false

        return true
    }

    override fun hashCode(): Int {
        var result = xmlns.contentHashCode()
        result = 31 * result + (description?.hashCode() ?: 0)
        result = 31 * result + binaries.hashCode()
        result = 31 * result + encoding.hashCode()
        return result
    }

    val title: String
        get() = description!!.titleInfo!!.bookTitle!!
    val lang: String?
        get() = description!!.titleInfo?.lang
    val annotation: Annotation?
        get() = description!!.titleInfo?.annotation

    companion object {
        val FB2_LOGGER: TaggedLogger = Logger.tag("FB2")
    }

}
